### PR TITLE
Amqp reusable connections 3

### PR DESCRIPTION
**Pull request by juanjoDiaz**
The third time lucky!

Continuation of #670

Everything as it was. And also includes the functionality wanted by #703 (connection provider based on a raw ConnectionFactory) but well documented and tested.

### CODE REVIEW

File: docs/src/main/paradox/amqp.md [hunk 1]
```diff
@@ -2,12 +2,7 @@
 
 The AMQP connector provides Akka Stream sources and sinks to connect to AMQP servers.
 
-### Reported issues
-
-[Tagged issues at Github](https://github.com/akka/alpakka/labels/p%3Aamqp)
```
Comment (ennru) [inline docs/src/main/paradox/amqp.md]: Please keep the reported issues section.

Comment (juanjoDiaz) [reply docs/src/main/paradox/amqp.md]: Done!

File: amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+          if (!state.compareAndSet(Closing, Empty))
+            throw new ConcurrentModificationException(
+              "Unexpected concurrent modification while closing the connection."
+            )
+        }
+      } else {
+        if (!state.compareAndSet(c, Connected(cachedConnection, clients - 1))) release(connection)
+      }
+    case Closing => release(connection)
+    case unknownState => throw new IllegalStateException(s"Unknown provider state ($unknownState).")
```
Comment (johanandren) [inline amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala]: Ah, now I see why you wrote "unknown" here, as state is sealed we don't actually have any unknown states, we only need to cover all the possible ones in the ADT, so you can remove the `catch whatever =>` part from both open and release

**Comment by juanjoDiaz** [comment]
Should be good now.

Just for my own understanding @johanandren, could you briefly explain what's the advantage of non-blocking CAS vs synchronized, for example? Aren't we blocking the thread anyway while the CAS is recursing?

**Comment by ennru** [comment]
Compilation errors.
And you'll need to rebase and run `test:compile` to get this year's header edition.

**Comment by ennru** [comment]
@juanjoDiaz There is more to the compilation errors. Classes define twice.
I'll wait for you to fix it before I cut 0.16 (hopefully today).

**Comment by juanjoDiaz** [comment]
Alright. It should be good now!

IntelliJ IDE must do some weird caching and refuses to use the 2018 headers :)

**Comment by ennru** [comment]
Cheers @juanjoDiaz you really kept it rolling here!

**Pull request merged**